File: sdkwork-models-python/sdkwork_models/loaders.py
from __future__ import annotations

import json
import os
from pathlib import Path
from urllib.parse import urljoin, urlparse
from urllib.request import urlopen

from .types import ModelCatalog


def _is_remote_url(value: str | Path) -> bool:
    parsed = urlparse(str(value))
    return parsed.scheme in {"http", "https"}


def _read_json(root: str | Path, rel_path: str) -> dict:
    if _is_remote_url(root):
        with urlopen(urljoin(f"{str(root).rstrip('/')}/", rel_path), timeout=30) as response:
            return json.loads(response.read().decode("utf-8"))
    return json.loads((Path(root) / rel_path).read_text(encoding="utf-8"))


def load_catalog(path_or_url: str | Path) -> ModelCatalog:
    manifest = _read_json(path_or_url, "sdkwork-models.json")
    meters = _read_json(path_or_url, "models/meters.json")["meters"]
    protocols = _read_json(path_or_url, "models/protocols.json")["protocols"]
    index = _read_json(path_or_url, "models/index.json")
    vendors: list[dict] = []
    vendor_codes: set[str] = set()
    vendor_catalogs: list[dict] = []
    models: list[dict] = []
    pricing: list[dict] = []
    for item in index.get("vendors", []):
        vendor_code = item["vendorCode"]
        region_code = item["regionCode"]
        vendor_catalog = load_vendor_catalog(path_or_url, vendor_code, region_code)
        vendor_catalogs.append(vendor_catalog)
        if vendor_code not in vendor_codes:
            vendor_codes.add(vendor_code)
            vendors.append(vendor_catalog["vendor"])
        models.extend(vendor_catalog["models"])
        pricing.extend(vendor_catalog["pricing"])
    return ModelCatalog(
        catalog_version=manifest["catalogVersion"],
        schema_version=manifest["schemaVersion"],
        meters=meters,
        protocols=protocols,
        vendors=vendors,
        vendor_catalogs=vendor_catalogs,
        models=models,
        pricing=pricing,
    )


def load_bundled_catalog() -> ModelCatalog:
    configured_root = os.getenv("SDKWORK_MODELS_CATALOG_ROOT", "").strip()
    if configured_root:
        return load_catalog(configured_root)
    return load_catalog(Path("data") / "sdkwork-models")


def load_vendor_catalog(path_or_url: str | Path, vendor_code: str, region_code: str) -> dict:
    index = _read_json(path_or_url, "models/index.json")
    vendor_index = next(
        (
            item
            for item in index.get("vendors", [])
            if item.get("vendorCode") == vendor_code and item.get("regionCode") == region_code
        ),
        None,
    )
    if vendor_index is None:
        raise ValueError(f"vendor region {vendor_code}/{region_code} is not indexed")
    return {
        "vendorCode": vendor_code,
        "regionCode": region_code,
        "vendor": _read_json(path_or_url, f"models/{vendor_index['path']}"),
        "families": _read_json(path_or_url, f"models/{vendor_index['familiesPath']}"),
        "models": [_read_json(path_or_url, f"models/{path}") for path in vendor_index.get("modelFiles", [])],
        "pricing": [_read_json(path_or_url, f"models/{path}") for path in vendor_index.get("pricingFiles", [])],
    }

File: sdkwork-models-flutter/lib/src/loaders.dart
import 'dart:convert';
import 'dart:io';

import 'types.dart';

Future<ModelCatalog> loadCatalog(String pathOrUrl) async {
  final manifest = await _readJsonObject(pathOrUrl, 'sdkwork-models.json');
  final meterFile = await _readJsonObject(pathOrUrl, 'models/meters.json');
  final protocolFile =
      await _readJsonObject(pathOrUrl, 'models/protocols.json');
  final index = await _readJsonObject(pathOrUrl, 'models/index.json');
  final vendors = <JsonObject>[];
  final seenVendorCodes = <String>{};
  final modelScores = <String, int>{};
  final vendorCatalogs = <JsonObject>[];
  final models = <String, JsonObject>{};
  final pricing = <JsonObject>[];
  for (final vendorIndex in _objectList(index['vendors'])) {
    final vendorCode = vendorIndex['vendorCode'];
    final regionCode = vendorIndex['regionCode'];
    if (vendorCode is! String) {
      continue;
    }
    if (regionCode is! String) {
      continue;
    }
    final vendorCatalog =
        await loadVendorCatalog(pathOrUrl, vendorCode, regionCode);
    vendorCatalogs.add(vendorCatalog);
    final vendor = vendorCatalog['vendor'];
    if (seenVendorCodes.add(vendorCode) && vendor is JsonObject) {
      vendors.add(vendor);
    }
    for (final model in _objectList(vendorCatalog['models'])) {
      _putBestModelIdentity(models, modelScores, vendorCatalog, model);
    }
    pricing.addAll(_objectList(vendorCatalog['pricing']));
  }
  return ModelCatalog(
    catalogVersion:
        _requiredString(manifest['catalogVersion'], 'catalogVersion'),
    schemaVersion: _requiredString(manifest['schemaVersion'], 'schemaVersion'),
    meters: _objectList(meterFile['meters']),
    protocols: _objectList(protocolFile['protocols']),
    vendors: vendors,
    vendorCatalogs: vendorCatalogs,
    models: models.values.toList(),
    pricing: pricing,
  );
}

Future<ModelCatalog> loadBundledCatalog() async {
  const envRoot = String.fromEnvironment('SDKWORK_MODELS_CATALOG_ROOT');
  if (envRoot.isNotEmpty) {
    return loadCatalog(envRoot);
  }
  final processRoot = Platform.environment['SDKWORK_MODELS_CATALOG_ROOT'];
  if (processRoot != null && processRoot.trim().isNotEmpty) {
    return loadCatalog(processRoot);
  }
  return loadCatalog('data/sdkwork-models');
}

Future<Map<String, Object?>> loadVendorCatalog(
    String pathOrUrl, String vendorCode, String regionCode) async {
  final index = await _readJsonObject(pathOrUrl, 'models/index.json');
  final vendorIndex =
      _objectList(index['vendors']).cast<JsonObject?>().firstWhere(
            (item) =>
                item?['vendorCode'] == vendorCode &&
                item?['regionCode'] == regionCode,
            orElse: () => null,
          );
  if (vendorIndex == null) {
    throw StateError('vendor region $vendorCode/$regionCode is not indexed');
  }
  return {
    'vendorCode': vendorCode,
    'regionCode': regionCode,
    'vendor': await _readJsonObject(pathOrUrl, 'models/${vendorIndex['path']}'),
    'families': await _readJsonObject(
        pathOrUrl, 'models/${vendorIndex['familiesPath']}'),
    'models': await _readJsonObjectsByRef(
        pathOrUrl, _stringList(vendorIndex['modelFiles'])),
    'pricing': await _readJsonObjectsByRef(
        pathOrUrl, _stringList(vendorIndex['pricingFiles'])),
  };
}

bool _isRemoteUrl(String value) {
  final uri = Uri.tryParse(value);
  return uri != null && (uri.scheme == 'http' || uri.scheme == 'https');
}

Future<JsonObject> _readJsonObject(String root, String relPath) async {
  final decoded = jsonDecode(await _readText(root, relPath));
  if (decoded is Map<String, Object?>) {
    return decoded;
  }
  if (decoded is Map) {
    return decoded.map((key, value) => MapEntry(key.toString(), value));
  }
  throw FormatException('JSON root must be an object: $relPath');
}

Future<String> _readText(String root, String relPath) async {
  if (_isRemoteUrl(root)) {
    final response = await HttpClient()
        .getUrl(Uri.parse('${root.replaceFirst(RegExp(r'/+$'), '')}/$relPath'))
        .then((request) => request.close());
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw HttpException(
          'failed to fetch sdkwork-models catalog file $relPath: ${response.statusCode}');
    }
    return utf8.decode(await response.expand((chunk) => chunk).toList());
  }
  return File('${Directory(root).path}/$relPath').readAsString();
}

Future<List<JsonObject>> _readJsonObjectsByRef(
    String root, List<String> refs) async {
  final result = <JsonObject>[];
  for (final ref in refs) {
    result.add(await _readJsonObject(root, 'models/$ref'));
  }
  return result;
}

List<JsonObject> _objectList(Object? value) {
  if (value is! List) {
    return const [];
  }
  return value
      .whereType<Map>()
      .map((item) => item.map((key, value) => MapEntry(key.toString(), value)))
      .toList();
}

List<String> _stringList(Object? value) {
  if (value is! List) {
    return const [];
  }
  return value.whereType<String>().toList();
}

String _requiredString(Object? value, String field) {
  if (value is String && value.isNotEmpty) {
    return value;
  }
  throw FormatException('$field must be a non-empty string');
}

void _putBestModelIdentity(
  Map<String, JsonObject> models,
  Map<String, int> scores,
  JsonObject vendorCatalog,
  JsonObject model,
) {
  final catalogKey = model['catalogKey'];
  if (catalogKey is! String || catalogKey.isEmpty) {
    return;
  }
  final score = _modelIdentityScore(vendorCatalog, model);
  final existingScore = scores[catalogKey];
  if (existingScore == null || score > existingScore) {
    models[catalogKey] = model;
    scores[catalogKey] = score;
  }
}

int _modelIdentityScore(JsonObject vendorCatalog, JsonObject model) {
  var score = 0;
  if (_hasRegionPricing(vendorCatalog, model)) {
    score += 100;
  }
  if (model['routingState'] == 'enabled') {
    score += 40;
  }
  if (model['shelfState'] == 'listed') {
    score += 20;
  }
  if (model['releaseStage'] == 'active') {
    score += 10;
  }
  if (model['lifecycle'] == 'current' || model['lifecycle'] == 'preview') {
    score += 5;
  }
  if (model['regionCode'] == 'global') {
    score += 1;
  }
  return score;
}

bool _hasRegionPricing(JsonObject vendorCatalog, JsonObject model) {
  final modelId = model['modelId'];
  if (modelId is! String) {
    return false;
  }
  return _objectList(vendorCatalog['pricing']).any((item) {
    return item['modelId'] == modelId && _objectList(item['prices']).isNotEmpty;
  });
}
